Add netgear plugin

DsgnrFH opened the pull request:
Squashed commits from main and from before

DsgnrFH committed:
Add netgear plugin

yhabteab commented:
Sorry @oxzi, it's my fault :)! After force purshing to the main branch GitHub automatically closed https://github.com/Icinga/check-netgear/pull/4 and can't also be reopened.

oxzi commented:
> Sorry @oxzi, it's my fault :)! After force purshing to the main branch GitHub automatically closed #4 and can't also be reopened.

Happens, I guess (:

@DsgnrFH: By having a new PR, all my old comments are now kinda void. Could you nevertheless comment under each of my comments in the other PR #4 if you have addressed it? I am going to take a closer look at this PR within this week, most likely tomorrow.

DsgnrFH commented:
> > Sorry @oxzi, it's my fault :)! After force purshing to the main branch GitHub automatically closed #4 and can't also be reopened.
> 
> Happens, I guess (:
> 
> @DsgnrFH: By having a new PR, all my old comments are now kinda void. Could you nevertheless comment under each of my comments in the other PR #4 if you have addressed it? I am going to take a closer look at this PR within this week, most likely tomorrow.

I believe I have addressed everything you suggested

yhabteab commented on internal/checks/checks.go:
```diff
@@ -0,0 +1,171 @@
+
+	"github.com/icinga/check-netgear/internal/utils"
+	"github.com/icinga/check-netgear/netgear"
+
```
redundant empty line.

yhabteab commented on internal/checks/checks.go:
```diff
@@ -0,0 +1,171 @@
+	"github.com/NETWAYS/go-check/result"
+)
+
+// CPU
```
That's not how you [document](https://go.dev/doc/comment#func) functions in Go.

yhabteab commented on internal/checks/checks.go:
```diff
@@ -0,0 +1,171 @@
+	partial := result.PartialResult{
+		Output: fmt.Sprintf("CPU Usage: %.2f%%", cpuUsage),
+	}
+	_ = partial.SetState(status)
```
Please make it a habit to not ignore errors like this. Instead, change the function return type like this because you don't have to return the `status` since you can access it via the `partial.GetStatus()` method if you need it.
```go
func CheckCPU(cpuUsage float64, warn float64, crit float64) (*result.PartialResult, error) {
	// ...
	if err := partial.SetState(status); err != nil {
		return nil, err
	}
	// ...
	return &partial, nil
}
```

The `//...` means the other stuff in that function that don't need to be changed.

Please adjust all the functions in this file just like this as well.

yhabteab commented on internal/checks/checks.go:
```diff
@@ -0,0 +1,171 @@
+		in := inRows[i]
+		out := outRows[i]
+
+		if !contains(portsToCheck, in.Port) {
```
You can use `!slices.Contains(portsToCheck, in.Port)` from the standard package instead and remove your `contains` function entirely.

yhabteab commented on netgear/netgear.go:
```diff
@@ -0,0 +1,165 @@
+	u, _ := url.Parse(strings.TrimSpace(hostName))
+	if u == nil {
+		u = &url.URL{}
+	}
```
Same here! If the user provides a URL that can't even be parsed why ignore it silently? This function should also return a `(*Netgear, err)` and you should return the error returned from `url.Parse` instead of ignoring it.

yhabteab commented on netgear/netgear.go:
```diff
@@ -0,0 +1,165 @@
+
+	return &Netgear{
+		client:   &http.Client{Timeout: timeout},
+		baseUrl:  *u,
```
You can also change the type of `baseUrl` to be `*url.URL` and you don't have to dereference it here with `*u`.

yhabteab commented on netgear/netgear.go:
```diff
@@ -0,0 +1,165 @@
+	return n.doRequest(http.MethodGet, "swcfg_poe", nil)
+}
+
+func (n *Netgear) doRequest(method, path string, body io.Reader) ([]byte, error) {
```
You don't seem to ever provide any request payload (something as a request body), since you're always passing `nil` for this `io.Reader`. Thus please remove the parameter from here and `doRequestURL` and pass `nil` directly to `http.NewRequest`.

yhabteab commented on main.go:
```diff
@@ -0,0 +1,181 @@
+	flag.Var(&mode, "mode", "Output modes to enable {basic|ports|poe|all} (repeatable)")
+
+	baseURL := flag.String("base-url", "http://192.168.1.1", "Base URL to use")
+	flag.StringVar(baseURL, "H", "http://192.168.1.1", "Base URL to use (shorthand)")
```
Alvar already pointed out that these default values look very strange in https://github.com/Icinga/check-netgear/pull/4#discussion_r2481894079.
> Furthermore, what's about the default value? This does not really look like the manufacturer's default?

Also, with the `flag` package from the standard library you can't use the same variable for multiple arguments. Is the shorthand argument really required? For now, I would suggest to drop all the shorthand arguments and later at some point, you can add these shorthand forms by using our Icinga Go Library instead but for now you don't necessarily need them.

DsgnrFH replied:
I believe the manufacturer's default was 192.168.0.239

yhabteab replied:
But where does the default IP you're using `192.168.1.1` come from? If there's a manufacturer's default IP for all devices, then you might use that as a default value instead. Otherwise, make this a required flag and don't set any default values.

DsgnrFH replied:
I have switched to using 192.168.0.239

yhabteab replied:
Sorry, I missed that.

yhabteab commented on netgear/types.go:
As Alvar suggested in https://github.com/Icinga/check-netgear/pull/4#discussion_r2481900343, you can add a brief [type description like this](https://go.dev/doc/comment#type) describing what the individual type represent in this file.

DsgnrFH committed:
Addressed some more comments and advises

yhabteab commented on internal/checks/checks.go:
```diff
@@ -0,0 +1,177 @@
+		partial.AddSubcheck(sub)
+	}
+
+	_ = partial.SetState(worst)
```
You're still ignoring errors here :)

yhabteab commented on internal/checks/checks.go:
```diff
@@ -0,0 +1,177 @@
+			sub := result.PartialResult{
+				Output: fmt.Sprintf("%s: %.2f%% loss", label, loss),
+			}
+			_ = sub.SetState(status)
```
Same here! You can change `addPerfSubcheck` like this: `addPerfSubcheck := func(label string, loss float64, status int) error {` and you can just do `return sub.SetState(status)` at the end of this function. And you can handle the error like this when calling it:
```go
		if err := addPerfSubcheck("IN", inLoss, inStatus); err != nil {
			return nil, err
		}
```

yhabteab commented on internal/checks/checks.go:
```diff
@@ -0,0 +1,177 @@
+				port.Port, state, port.CurrentPower/1000, port.PowerLimit/1000,
+			),
+		}
+		_ = poeCheck.SetState(status)
```
Still ignoring an error.

yhabteab commented on netgear/netgear.go:
```diff
@@ -0,0 +1,168 @@
+	}
+	if u == nil {
+		u = &url.URL{}
+	}
```
Since you're correctly handling the parsing error now, `u == nil` will never be true, so you can remove this. Just make sure in the main function that the `hostName` (should probably be renamed to `url` or something) begins with `http://` or `https://`.

yhabteab commented on main.go:
```diff
@@ -0,0 +1,206 @@
+	n, err := netgear.NewNetgear(*baseURL, *username, *password)
+	if err != nil {
+		fmt.Printf("URL error: %v", err)
+		os.Exit(1)
```
You can use `check.Unknown` here as well.

yhabteab commented on main.go:
```diff
@@ -0,0 +1,206 @@
+		if !*hidecpu {
+			cpuPartial, err := checks.CheckCPU(cpuUsage, *cpuWarn, *cpuCrit)
+			if err != nil {
+				fmt.Printf("Error checking CPU: %v\n", err)
```
I think adding this error case as a subcheck like this would be better.
```go
				errRes := result.NewPartialResult()
				errRes.Output = fmt.Sprintf("CPU check error: %v", err)
				o.AddSubcheck(errRes)
```

And same for all the other cases below (`memory`, `temperature`, fans, ports, poe etc.).

DsgnrFH committed:
Addressed even more suggestions

DsgnrFH committed:
Fixed main

DsgnrFH committed:
Removed unused import

DsgnrFH committed:
Addressed more comments

yhabteab commented on internal/checks/checks.go:
```diff
@@ -0,0 +1,184 @@
+		}
+
+		addPerfSubcheck("IN", inLoss, inStatus)
+		addPerfSubcheck("OUT", outLoss, outStatus)
```
```suggestion
		if err := addPerfSubcheck("IN", inLoss, inStatus); err != nil {
			return nil, err
		}
		if err := addPerfSubcheck("OUT", outLoss, outStatus); err != nil {
			return nil, err
		}
```

yhabteab commented on main.go:
```diff
@@ -0,0 +1,227 @@
+	hidetemp := flag.Bool("notemp", false, "Hide the Temperature info")
+	hidefans := flag.Bool("nofans", false, "Hide the Fans info")
+
+	var mode stringSliceFlag = []string{"basic"}
```
```suggestion
	mode := stringSliceFlag{"basic"}
```

yhabteab commented on main.go:
```diff
@@ -0,0 +1,227 @@
+	statsWarn := flag.Float64("stats-warning", 5, "Port stats warning threshold")
+	statsCrit := flag.Float64("stats-critical", 20, "Port stats critical threshold")
+
+	var portsToCheck intSliceFlag = []int{1, 2, 3, 4, 5, 6, 7, 8}
```
```suggestion
	portsToCheck := intSliceFlag{1, 2, 3, 4, 5, 6, 7, 8}
```

yhabteab commented on main.go:
```diff
@@ -0,0 +1,227 @@
+
+		upTime := deviceInfo.DeviceInfo.Details[0].Uptime
+		cpuUsage, _ := netgear.StringPercentToFloat(deviceInfo.DeviceInfo.Cpu[0].Usage)
+		memUsage, _ := netgear.StringPercentToFloat(deviceInfo.DeviceInfo.Memory[0].Usage)
```
You're still ignoring errors.

yhabteab reviewed (commented):
@oxzi these are my last review comments, so you can also have look at this, I won't intervene further :-)

oxzi commented on main.go:
```diff
@@ -0,0 +1,227 @@
+	if *help || *username == "" || *password == "" {
+		flag.Usage()
+		return
+	}
```
I would split these up in two cases. If someone request the help, it's okay. However, if someone forgets to set credentials - or they somehow get lost along the way -, it is an error.

```suggestion
	if *help {
		flag.Usage()
		return
	}

	if *username == "" || *password == "" {
		fmt.Println("Both username and password are required")
		os.Exit(check.Unknown)
	}
```

oxzi commented on main.go:
```diff
@@ -0,0 +1,227 @@
+		fan := deviceInfo.DeviceInfo.Fan[0].Details[0]
+		fanName := fan.Description
+		fanSpeed := fan.Speed
+		sensorDetails := deviceInfo.DeviceInfo.Sensor[0].Details
```
Are you sure that there is always an element at index `0`? What if the API slightly changes or something along the way has a bad day.

Please consider checking if the slice is long enough before accessing something. Otherwise, if you say that this _MUST_ work, consider adding a deferred `recover()` call, which prints the error and exits with `UNKNOWN`.

oxzi commented on main.go:
```diff
@@ -0,0 +1,227 @@
+			tempPartial, err := checks.CheckTemperature(sensorDetails, *tempWarn, *tempCrit)
+			if err != nil {
+				errRes := result.NewPartialResult()
+				errRes.Output = fmt.Sprintf("Temperatuer check error: %v", err)
```
```suggestion
				errRes.Output = fmt.Sprintf("Temperature check error: %v", err)
```

oxzi commented on netgear/netgear.go:
```diff
@@ -0,0 +1,174 @@
+		return nil, err
+	}
+	return &di, nil
+}
+
+func (n *Netgear) PortStatistics(statType string) (*PortStatistics, error) {
+	const defaultPageIndex = 1
+	const defaultPageSize = 25
+
+	u := n.baseUrl.JoinPath("port_statistics")
+	q := u.Query()
+	q.Set("type", statType)
+	q.Set("indexPage", strconv.Itoa(defaultPageIndex))
+	q.Set("pageSize", strconv.Itoa(defaultPageSize))
```
As [commented in the old PR](https://github.com/Icinga/check-netgear/pull/4#discussion_r2481848798), this pagination looks a bit weird.

What about ports on page 2? Can there be a second page? Does this mean we are only getting information about 25 ports? If so, lots of switches would miss information.

TheSyscall replied:
That is true, but the API currently always returns the first page and the "pageSize" parameter seems to have no effect.

Adding a comment for that known bug.

oxzi commented on netgear/netgear.go:
```diff
@@ -0,0 +1,174 @@
+}
+
+func NewNetgear(hostName, username, password string) (*Netgear, error) {
+	u, err := url.Parse(strings.TrimSpace(hostName))
```
https://github.com/Icinga/check-netgear/pull/4#discussion_r2481808159

```suggestion
func NewNetgear(baseUrl, username, password string) (*Netgear, error) {
	u, err := url.Parse(strings.TrimSpace(baseUrl))
```

TheSyscall committed:
change types to reflect what is actually sent by the switch

TheSyscall committed:
add error handling

TheSyscall committed:
move getting of values to where we know that we actually need them

TheSyscall committed:
Use "basic" mode as default, but allow others to override it

TheSyscall committed:
explicitly require username and password

TheSyscall committed:
fix typo "Temperatuer"

TheSyscall committed:
rename "hostName" to "baseUrl"

TheSyscall committed:
move modes into their own functions

TheSyscall committed:
check array length before access

TheSyscall committed:
add notice for missing pagination

TheSyscall committed:
combine all checks into a single overall output

This both allows multiple modes to be active at the same time, and
causes the perfdata to be combined at the end, as it should be.

TheSyscall committed:
make perfdata optional

TheSyscall committed:
show an error when no mode was executed

oxzi commented on main.go:
```diff
@@ -0,0 +1,364 @@
+func ModeBasic(netgearSession *netgear.Netgear, worstStatus *int, flags *Flags) (*result.PartialResult, error) {
+	deviceInfo, err := netgearSession.DeviceInfo()
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving device info: %v\n", err)
```
1. An error should not end in a newline or another white space.
2. When embedding one error into another, please use [error wrapping](https://go.dev/blog/go1.13-errors#errors-in-go-113). For `fmt.Errorf`, use the `%w` verb.


```suggestion
		return nil, fmt.Errorf("error retrieving device info: %w", err)
```

Same applies to line 90 and 114.

oxzi commented on main.go:
```diff
@@ -0,0 +1,364 @@
+	flag.BoolVar(&flags.HideFans, "nofans", false, "Hide the Fans info")
+
+	mode := stringSliceFlag{}
+	flag.Var(&mode, "mode", "Output modes to enable {basic|ports|poe|all} (repeatable)")
```
https://github.com/Icinga/check-netgear/blob/cacd4debed546b0dbc479538890d9bc6490da978/main.go#L317-L318

When not set, `mode` falls back to `basic`. This should be documented in the `-h` output text.

oxzi commented on main.go:
```diff
@@ -0,0 +1,364 @@
+
+// ModeBasic contains all the basic hardware information of the switch, including CPU and RAM usage, temperature and fan
+// speed
+func ModeBasic(netgearSession *netgear.Netgear, worstStatus *int, flags *Flags) (*result.PartialResult, error) {
```
Having `worstStatus` as an int pointer argument to be altered within the function to be used as a return value feels very C-ish and is not idiomatic in Go.

In Go, you would just return multiple values, one would be the state for this check. Later, you can just use `worstStatus = max(worstStatus, checkStatus)` to update the worst state.

Same applies to the other two `ModePorts` and `ModePoE` functions.

oxzi commented on main.go:
```diff
@@ -0,0 +1,364 @@
+	upTime := deviceInfo.DeviceInfo.Details[0].Uptime
+	if len(deviceInfo.DeviceInfo.Details) == 0 {
+		return nil, fmt.Errorf("error retrieving device info")
+	}
```
The `if` can never be true. If there are no `.Details`, the access to `.Details[0].Uptime` will already panic. These two operations should be switched.


```suggestion
	if len(deviceInfo.DeviceInfo.Details) == 0 {
		return nil, fmt.Errorf("error retrieving device info")
	}
	upTime := deviceInfo.DeviceInfo.Details[0].Uptime
```

oxzi commented on netgear/netgear.go:
```diff
@@ -0,0 +1,177 @@
+	if err != nil {
+		return err
+	}
+	defer func() { _ = resp.Body.Close() }()
```
> If the Body is not both read to EOF and closed, the [Client](https://pkg.go.dev/net/http#Client)'s underlying [RoundTripper](https://pkg.go.dev/net/http#RoundTripper) (typically [Transport](https://pkg.go.dev/net/http#Transport)) may not be able to re-use a persistent TCP connection to the server for a subsequent "keep-alive" request.

-- <https://pkg.go.dev/net/http#Client.Do>

---

Your `json.NewDecoder` might fail for unknown input or whatnot. Just make sure to always drain the response body.


```suggestion
	defer func() {
		_, _ = io.Copy(io.Discard, resp.Body)
		_ = resp.Body.Close()
	}()
```

oxzi commented on netgear/netgear.go:
```diff
@@ -0,0 +1,177 @@
+		return err
+	}
+	defer func() { _ = resp.Body.Close() }()
+	_, _ = io.Copy(io.Discard, resp.Body)
```
No reason to `defer` here.

```suggestion
	_, _ = io.Copy(io.Discard, resp.Body)
	_ = resp.Body.Close()
```

oxzi commented on netgear/netgear.go:
```diff
@@ -0,0 +1,177 @@
+	if err := n.doRequest(http.MethodGet, "swcfg_poe", &poeStatus); err != nil {
+		return nil, err
+	}
+	return &poeStatus, nil
```
You can skip some references by directly starting with a pointer.


```suggestion
	poeStatus := new(PoeStatus)
	if err := n.doRequest(http.MethodGet, "swcfg_poe", poeStatus); err != nil {
		return nil, err
	}
	return poeStatus, nil
```

oxzi commented on netgear/netgear.go:
```diff
@@ -0,0 +1,177 @@
+		// ignore response body but still read it so connection can be reused
+		_, _ = io.Copy(io.Discard, resp.Body)
+		return nil
+	}
```
1. The body draining can be moved to the deferred function regardless of the result.
2. Why is a nil pointer for `result` valid? I have found no `doRequestURL` or `doRequest` method call with a nil pointer, indicating the result should be discarded. Thus, either let it crash or document why it is required.

TheSyscall replied:
2. While there is no place where we call `doRequestURL` without a result object, I still see value in allowing HTTP requests without requiring a json formatted body back.

oxzi replied:
Totally fine, but then it should be documented in the method doc comment. Otherwise, it would be unspecified/unexpected behavior.

TheSyscall committed:
Apply suggestions from code review

Co-authored-by: Alvar <github.jyf300hj@0x21.biz>

oxzi commented on main.go:
```diff
@@ -0,0 +1,362 @@
+		}
+		cpuUsage, err := netgear.StringPercentToFloat(deviceInfo.DeviceInfo.Cpu[0].Usage)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing CPU usage: %v\n", err)
```
Still needs to be addressed, as I stated [yesterday](https://github.com/Icinga/check-netgear/pull/6#discussion_r2560006050).


```suggestion
			return nil, fmt.Errorf("error parsing CPU usage: %w", err)
```

TheSyscall replied:
Changed it.

oxzi commented on main.go:
```diff
@@ -0,0 +1,362 @@
+		}
+		memUsage, err := netgear.StringPercentToFloat(deviceInfo.DeviceInfo.Memory[0].Usage)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing Memory usage: %v\n", err)
```
https://github.com/Icinga/check-netgear/pull/6#discussion_r2560006050

TheSyscall replied:
Changed it

oxzi reviewed (commented):
For future reviews: Please do not close the review comments by yourself, but post there briefly what you have changed. If it is in a distinct commit, just reference the commit hash.

Doing so eases the process for the reviewer by a lot. At the moment, I have to manually check how you have addressed each review, and do a local git diff between my local last state and what origin now has. GitHub is not really helpful in this regard.

So please interact with the comments.

TheSyscall commented:
> For future reviews: Please do not close the review comments by yourself, but post there briefly what you have changed. If it is in a distinct commit, just reference the commit hash.
> 
> Doing so eases the process for the reviewer by a lot. At the moment, I have to manually check how you have addressed each review, and do a local git diff between my local last state and what origin now has. GitHub is not really helpful in this regard.
> 
> So please interact with the comments.

Sorry, will do that from now on.

oxzi reviewed (approved):
Thanks, looks good so far!

There are still some points to iterate on, but I would say it's enough for this PR not to result in bikeshedding.

For the future, I would love to see some mocked API tests. At the moment, the check plugin relies on API responses either from the OpenAPI spec or from your experience. Since I don't have such a Netgear Switch at hand, I cannot verity that this check command actually works. Thus, please consider to add tests with actual responses from the device's API in a future PR.

Thanks for all your work, to both of you!
